### PR TITLE
Server configuration variables & HTTP/HTTPS/S3 file system adapter

### PR DESCRIPTION
**Pull request by hannes**
This PR contains two parts, the server configuration variables interface (1) and a HTTP/HTTPS/S3 file system adapter (2)

### (1) Variable Iterface

Postgres can set some server configuration variables using the `SET a = 42` syntax. We hijack this syntax to store arbitrary key/value pairs in the running database. Please note that these settings are currently **global**, **immediate** (non-transactional) and **transient** (not persistent). 

That being said, set a variable like so
````SQL
SET my_variable = 'some_value';
SET other_variable = 42;
````
Variables can be overwritten.

You can retrieve the current value of a variable within queries using the `CURRENT_SETTING` scalar function, like so:
````SQL
SELECT CURRENT_SETTING('my_variable');
````
This can be very useful for example when using values in multiple places within a SQL script, or even to pass settings to the shell like so for example:
````bash
duckdb -cmd "set some_variable='some_value'" < script.sql
````
with `script.sql` then containing `CURRENT_SETTING('some_variable')` calls. 

### (2) HTTP/HTTPS/S3 file system adapter
Even though we do not believe that the cloud is a real thing, we are still interested in making use of the allegedly large computing resources available there. We have heard that data lives in a lake called "S3" nowadays. In order to make it possible to read Parquet files that are stored on S3 or really any HTTP/HTTPs server, we add the `httpfs` extension. This extension hacks our file system interface to allow special protocol handlers for special cases such as HTTP. 

The extension is enabled for source builds by passing the `BUILD_HTTPFS=1` to `make` or `-DBUILD_HTTPFS_EXTENSION=1` to `cmake`. Once enabled and loaded, the extension can be used transparently like so:
````SQL
SELECT * FROM PARQUET_SCAN('https://github.com/cwida/duckdb/blob/master/test/sql/copy/parquet/data/userdata1.parquet?raw=true');
````

The S3 extension needs some configuration, since its likely that you need to set access credentials for it. Remember those `SET` variables? They somehow come in handy here:
````SQL
SET s3_region='your-region';
SET s3_access_key_id='your_key_id';
SET s3_secret_access_key='your_secret_key';

SELECT * FROM PARQUET_SCAN('s3://my_bucket/my_file') LIMIT 10;
````

There are still some open issues for this S3 reader:
- Wildcards in file names are not supported yet. Yes, they should be.
- Caching is not yet optimised for the multi-cursor reads that are inherent when scanning Parquet files. Will come.
- We have not tried to optimise the amount of requests down. So if you use this and you get a `uuge bill, you have been warned.

### CODE REVIEW

File: extension/httpfs/httpfs.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	auto slash_pos = url.find('/', 8);
+	if (slash_pos == std::string::npos) {
+		throw std::runtime_error("URL needs to contain a '/' after the host");
+	}
+	auto proto_host_port = url.substr(0, slash_pos);
+
+	auto path = url.substr(slash_pos);
+	if (path.empty()) {
+		throw std::runtime_error("URL needs to contain a bucket name");
```
Comment (Mause) [inline extension/httpfs/httpfs.cpp:26]: Copied error message?

File: src/include/duckdb/common/file_system.hpp [hunk 2]
```diff
@@ -130,9 +132,121 @@ class FileSystem {
...
+	}
+
+	void RegisterProtocolHandler(string protocol, unique_ptr<FileSystem> protocol_fs) override {
+		protocol_handler_fss[protocol] = move(protocol_fs);
+	}
+
+private:
+	FileSystem *FindFileSystem(string path) {
+		for (auto &handler : protocol_handler_fss) {
+			if (path.rfind(handler.first, 0) == 0) {
```
Comment (Mytherin) [inline src/include/duckdb/common/file_system.hpp:240]: `memcmp` should work here as well no? Or am I missing something?

Comment (hannesmuehleisen) [reply src/include/duckdb/common/file_system.hpp:240]: This did not seem like something worth optimising given the low number of entries in map. `rfind` seems to be implemented fairly efficiently: https://github.com/llvm/llvm-project/blob/3696227c10f5e5841223c2a2fb63fdd1d50a7930/libcxx/include/__string#L1016

Comment (Mytherin) [reply src/include/duckdb/common/file_system.hpp:240]: Ah fair enough, indeed that seems like it essentially does a memcmp under the hood when `pos=0`.

**Pull request merged**